File: apps/backend/src/ditto_apps/models/technical_analysis.py
"""HTTP contracts for exact deterministic technical-analysis snapshots."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal, cast

from ditto_kernel.identity import InstrumentId
from pydantic import BaseModel, BeforeValidator, ConfigDict, Field

__all__ = [
    "TechnicalAnalysisConflictResponse",
    "TechnicalAnalysisIndicatorParameterResponse",
    "TechnicalAnalysisIndicatorResponse",
    "TechnicalAnalysisLevelResponse",
    "TechnicalAnalysisQueryBody",
    "TechnicalAnalysisSnapshotResponse",
    "TechnicalAnalysisSpecRequest",
    "TechnicalAnalysisTimeframeResponse",
]

_REQUEST_CONFIG = ConfigDict(strict=True, extra="forbid")
_RESPONSE_CONFIG = ConfigDict(strict=True, frozen=True, from_attributes=True)

type Timeframe = Literal["daily", "weekly"]
type Direction = Literal["bullish", "bearish", "neutral", "unknown"]


def _parse_http_datetime(value: object) -> object:
    if not isinstance(value, str):
        return value
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def _parse_http_array(value: object) -> object:
    if isinstance(value, list):
        return tuple(cast("list[object]", value))
    return value


HttpDateTime = Annotated[datetime, BeforeValidator(_parse_http_datetime)]
Timeframes = Annotated[tuple[Timeframe, ...], BeforeValidator(_parse_http_array)]
SnapshotIds = Annotated[tuple[str, ...], BeforeValidator(_parse_http_array)]


class TechnicalAnalysisSpecRequest(BaseModel):
    """Complete versioned v1 technical-indicator parameter set."""

    model_config = _REQUEST_CONFIG

    spec_id: str = Field(min_length=1)
    spec_version: str = Field(min_length=1)
    algorithm_version: Literal["technical-analysis.v1"] = "technical-analysis.v1"
    timeframes: Timeframes = Field(min_length=1)
    return_window: int = Field(default=20, gt=0)
    trend_window: int = Field(default=20, gt=0)
    slope_window: int = Field(default=5, ge=2)
    rsi_window: int = Field(default=14, gt=0)
    macd_fast: int = Field(default=12, gt=0)
    macd_slow: int = Field(default=26, gt=0)
    macd_signal: int = Field(default=9, gt=0)
    atr_window: int = Field(default=14, gt=0)
    volatility_window: int = Field(default=20, gt=0)
    volume_window: int = Field(default=20, gt=0)
    donchian_window: int = Field(default=20, gt=0)
    support_resistance_window: int = Field(default=60, gt=0)


class TechnicalAnalysisQueryBody(BaseModel):
    """One exact read request; source facts remain server-owned."""

    model_config = _REQUEST_CONFIG

    instrument_id: InstrumentId = Field(gt=0)
    instrument_name: str = Field(min_length=1)
    instrument_code: str = Field(min_length=1)
    as_of: HttpDateTime
    knowledge_cutoff: HttpDateTime
    publication_cutoff: HttpDateTime
    source_snapshot_ids: SnapshotIds = Field(min_length=1)
    spec: TechnicalAnalysisSpecRequest
    selection_run_id: str | None = None
    research_case_id: str | None = None
    portfolio_snapshot_id: str | None = None


class TechnicalAnalysisIndicatorParameterResponse(BaseModel):
    """One canonical numeric indicator parameter."""

    model_config = _RESPONSE_CONFIG

    name: str
    value: int


class TechnicalAnalysisIndicatorResponse(BaseModel):
    """One indicator result with explicit warm-up or unavailable state."""

    model_config = _RESPONSE_CONFIG

    name: str
    timeframe: Timeframe
    indicator_version: str
    window: int | None
    parameters: tuple[TechnicalAnalysisIndicatorParameterResponse, ...]
    value: float | None
    status: Literal["ready", "warming_up", "unavailable"]
    reason: str | None


class TechnicalAnalysisLevelResponse(BaseModel):
    """One versioned support or resistance level."""

    model_config = _RESPONSE_CONFIG

    timeframe: Timeframe
    kind: Literal["support", "resistance"]
    price: float
    confidence: float
    touches: int
    window: int
    algorithm_version: str


class TechnicalAnalysisTimeframeResponse(BaseModel):
    """Compact technical state for one timeframe."""

    model_config = _RESPONSE_CONFIG

    timeframe: Timeframe
    trend: Direction
    momentum: Direction
    breakout: Direction


class TechnicalAnalysisConflictResponse(BaseModel):
    """One deterministic daily/weekly disagreement."""

    model_config = _RESPONSE_CONFIG

    dimension: Literal["trend", "momentum", "breakout"]
    daily: Direction
    weekly: Direction
    reason_code: str


class TechnicalAnalysisSnapshotResponse(BaseModel):
    """Content-addressed exact technical-analysis result."""

    model_config = _RESPONSE_CONFIG

    snapshot_id: str
    input_hash: str
    spec_hash: str
    registry_version: str
    instrument_id: InstrumentId
    instrument_name: str
    as_of: datetime
    knowledge_cutoff: datetime
    publication_cutoff: datetime
    source_snapshot_ids: tuple[str, ...]
    status: Literal["ready", "degraded", "blocked"]
    last_visible_bar_at: datetime | None
    last_computed_bar_at: datetime | None
    readings: tuple[TechnicalAnalysisIndicatorResponse, ...]
    levels: tuple[TechnicalAnalysisLevelResponse, ...]
    timeframe_summaries: tuple[TechnicalAnalysisTimeframeResponse, ...]
    conflicts: tuple[TechnicalAnalysisConflictResponse, ...]
    missing_inputs: tuple[str, ...]
    warnings: tuple[str, ...]
    selection_run_id: str | None
    research_case_id: str | None
    portfolio_snapshot_id: str | None

File: apps/web/src/mocks/handlers/technical-analysis.ts
import { HttpResponse, http } from "msw";
import type {
	TechnicalAnalysisQueryBody as QueryBody,
	TechnicalAnalysisSnapshot as Snapshot,
} from "@/features/instruments/api/technical-analysis";

export const technicalAnalysisHandlers = [
	http.post("/api/v1/technical-analysis/snapshots/query", async ({ request }) => {
		const body = (await request.json()) as QueryBody;
		const snapshot = {
			as_of: body.as_of,
			conflicts: [
				{ daily: "bullish", dimension: "trend", reason_code: "daily_weekly_disagreement", weekly: "bearish" },
			],
			input_hash: "e".repeat(64),
			instrument_id: body.instrument_id,
			instrument_name: body.instrument_name,
			knowledge_cutoff: body.knowledge_cutoff,
			last_computed_bar_at: body.as_of,
			last_visible_bar_at: body.as_of,
			levels: [
				{
					algorithm_version: "support-resistance.v1",
					confidence: 0.82,
					kind: "support",
					price: 1718,
					timeframe: "daily",
					touches: 3,
					window: 60,
				},
				{
					algorithm_version: "support-resistance.v1",
					confidence: 0.74,
					kind: "resistance",
					price: 1808,
					timeframe: "daily",
					touches: 2,
					window: 60,
				},
			],
			missing_inputs: [],
			portfolio_snapshot_id: body.portfolio_snapshot_id ?? null,
			publication_cutoff: body.publication_cutoff,
			readings: [
				{
					indicator_version: "sma.v1",
					name: "sma",
					parameters: [{ name: "window", value: 20 }],
					reason: null,
					status: "ready",
					timeframe: "daily",
					value: 1742.35,
					window: 20,
				},
				{
					indicator_version: "rsi.v1",
					name: "rsi",
					parameters: [{ name: "window", value: 14 }],
					reason: null,
					status: "ready",
					timeframe: "weekly",
					value: 61.24,
					window: 14,
				},
			],
			registry_version: "technical-indicator-registry.v1",
			research_case_id: body.research_case_id ?? null,
			selection_run_id: body.selection_run_id ?? null,
			snapshot_id: `technical-analysis:sha256:${"f".repeat(64)}`,
			source_snapshot_ids: body.source_snapshot_ids,
			spec_hash: "a".repeat(64),
			status: "ready",
			timeframe_summaries: [
				{ breakout: "neutral", momentum: "bullish", timeframe: "daily", trend: "bullish" },
				{ breakout: "neutral", momentum: "neutral", timeframe: "weekly", trend: "bearish" },
			],
			warnings: [],
		} satisfies Snapshot;
		return HttpResponse.json({ data: snapshot });
	}),
];
